### DISCUSSION TITLE
Voronoi diagram optimized by Lloyd relaxation

### DISCUSSION THREAD

**Discussion by zyyue99** [BrunoLevy/geogram#195] [Q&A] [answered]
Hi, could you please tell me how to get a 3D Voronoi diagram optimized by Lloyd relaxation? What Delaunay3D seems to give is just a demo of working with a cube. Thank you very much.

**Comment by BrunoLevy** [comment]
Note sure that I understand what you need, do you mean you need to clip the Voronoi diagram by something else than a cube ? For that, take a look at the [compute_RVD](https://github.com/BrunoLevy/geogram/tree/main/src/examples/geogram/compute_RVD) example.

**Comment by zyyue99** [comment]
Thank you for your reply. What I need is a clipped Voronoi diagram. However, it seems that the compute_RVD does not optimize the pointset using Lloyd relaxation. I currently have an initial set of points, but I don't think it's distributed evenly enough. If I want to optimize it with Lloyd relaxation, are there any tools I can use?

**Comment by BrunoLevy** [comment, marked as answer]
You can find two examples of Lloyd relaxation / Centroidal Voronoi Tesselation with clipped volumes :
- `vorpalite` (bundled with geogram) polyhedral mesher [here](https://github.com/BrunoLevy/geogram/blob/1b92f86a38ad4bfe4e8e7f316293fc3672608676/src/bin/vorpalite/main.cpp#L358)
- `sample_volume` command of the Graphite software [here](https://github.com/BrunoLevy/GraphiteThree/blob/6b67b827bde2d549f5d99b62db0cf7dc33c71814/src/lib/OGF/mesh/commands/mesh_grob_points_commands.cpp#L287)
